### ISSUE TITLE
Autoscaling workers can brick pipelines

### ISSUE THREAD

**Issue by georgehaidar** [concourse/concourse#712]
# Bug Report

Following scale down events and scheduled shutdown of our ConcourseCI workers, we are finding that some pipeline resources and jobs get stuck in error state with `Worker ##### not found errors`. Attempting to nudge resource checks back with with `fly check-resource` does not work (the command just hangs). The only way around this problem has been to terminate the ATC and workers.

* Concourse version: 2.3.1
* Deployment type: Docker over AWS ECS

## Admin Log

```
{"timestamp":"1477020574.488196135","source":"atc","message":"atc.baggage-collector.could-not-locate-worker","log_level":1,"data":{"session":"16","worker-id":"c29f3c179037"}}
{"timestamp":"1477020574.488203049","source":"atc","message":"atc.baggage-collector.could-not-locate-worker","log_level":1,"data":{"session":"16","worker-id":"dd14d9a3c373"}}
{"timestamp":"1477020574.488210440","source":"atc","message":"atc.baggage-collector.could-not-locate-worker","log_level":1,"data":{"session":"16","worker-id":"d6c571c3448c"}}
{"timestamp":"1477020574.488217354","source":"atc","message":"atc.baggage-collector.could-not-locate-worker","log_level":1,"data":{"session":"16","worker-id":"575ed199ab86"}}
{"timestamp":"1477020574.488225222","source":"atc","message":"atc.baggage-collector.could-not-locate-worker","log_level":1,"data":{"session":"16","worker-id":"87dfb313e736"}}
{"timestamp":"1477020574.488241673","source":"atc","message":"atc.baggage-collector.could-not-locate-worker","log_level":1,"data":{"session":"16","worker-id":"b9218d7cb954"}}
{"timestamp":"1477020574.488453627","source":"atc","message":"atc.baggage-collector.retry-round-tripper.retrying","log_level":1,"data":{"error":"dial tcp 0.0.0.0:33534: getsockopt: connection refused","failed-attempts":1,"ran-for":"169.231µs","session":"16.1121"}}
{"timestamp":"1477020574.488233566","source":"atc","message":"atc.baggage-collector.could-not-locate-worker","log_level":1,"data":{"session":"16","worker-id":"4ed32bf48bbd"}}
{"timestamp":"1477020575.434213161","source":"atc","message":"atc.garden-connection.retryable-http-client.retrying","log_level":1,"data":{"error":"dial tcp 0.0.0.0:33647: getsockopt: connection refused","failed-attempts":3,"ran-for":"3.012680461s","session":"68192.1"}}
{"timestamp":"1477020575.488822222","source":"atc","message":"atc.baggage-collector.retry-round-tripper.retrying","log_level":1,"data":{"error":"dial tcp 0.0.0.0:33534: getsockopt: connection refused","failed-attempts":2,"ran-for":"1.000526233s","session":"16.1121"}}
{"timestamp":"1477020575.873320341","source":"atc","message":"atc.syncer.ato-bas:radar.scan.tick.init.retry-round-tripper.retrying","log_level":1,"data":{"container":"5a145d53-aa21-4226-4b57-ec6775e7a33e","error":"dial tcp 0.0.0.0:46511: getsockopt: connection refused","failed-attempts":133,"pipeline-scoped-name":"ato-bas:resource:activity-statement-service","ran-for":"34m23.056148171s","session":"13.9.124.46.1.11"}}
{"timestamp":"1477020576.293712139","source":"atc","message":"atc.syncer.taxonline:radar.scan.tick.init.retry-round-tripper.retrying","log_level":1,"data":{"container":"49102730-e857-4af9-404a-7b737d7eda24","error":"dial tcp 0.0.0.0:36740: getsockopt: connection refused","failed-attempts":89,"pipeline-scoped-name":"taxonline:resource:form-operations","ran-for":"22m39.039216459s","session":"13.1.218.202.1.10"}}
{"timestamp":"1477020576.456609249","source":"atc","message":"atc.syncer.taxonline:radar.scan.tick.init.retry-round-tripper.retrying","log_level":1,"data":{"container":"13d786c5-8b36-4973-67b3-4959f3af0b0e","error":"dial tcp 0.0.0.0:46511: getsockopt: connection refused","failed-attempts":133,"pipeline-scoped-name":"taxonline:resource:prod-po-assets","ran-for":"34m23.063215634s","session":"13.1.157.58.1.11"}}
{"timestamp":"1477020576.634534121","source":"atc","message":"atc.syncer.sdk:radar.scan.tick.init.retry-round-tripper.retrying","log_level":1,"data":{"container":"4f23a92f-607c-458b-4bac-1ff7500b0093","error":"dial tcp 0.0.0.0:36740: getsockopt: connection refused","failed-attempts":89,"pipeline-scoped-name":"sdk:resource:common","ran-for":"22m39.039258951s","session":"13.5.27.55.1.10"}}
{"timestamp":"1477020576.685718060","source":"tsa","message":"tsa.connection.keepalive","log_level":1,"data":{"remote":"10.0.84.97:44257","session":"15","type":"keepalive"}}
{"timestamp":"1477020577.165316820","source":"atc","message":"atc.syncer.taxonline:radar.scan.tick.init.retry-round-tripper.retrying","log_level":1,"data":{"container":"2e69d365-8ea0-4f44-60ca-3d819ca8a17d","error":"dial tcp 0.0.0.0:36740: getsockopt: connection refused","failed-attempts":89,"pipeline-scoped-name":"taxonline:resource:ato-fbt-2015","ran-for":"22m39.038508555s","session":"13.1.216.202.1.10"}}
{"timestamp":"1477020577.305937767","source":"atc","message":"atc.syncer.docker-images:radar.scan.tick.init.retry-round-tripper.retrying","log_level":1,"data":{"container":"6ead43af-a13a-4a68-5eba-c54593fb734b","error":"dial tcp 0.0.0.0:36740: getsockopt: connection refused","failed-attempts":89,"pipeline-scoped-name":"docker-images:resource:node-service-version","ran-for":"22m39.037524759s","session":"13.3.72.193.1.10"}}
{"timestamp":"1477020577.489269733","source":"atc","message":"atc.baggage-collector.retry-round-tripper.retrying","log_level":1,"data":{"error":"dial tcp 0.0.0.0:33534: getsockopt: connection refused","failed-attempts":3,"ran-for":"3.000983855s","session":"16.1121"}}
{"timestamp":"1477020577.970918417","source":"tsa","message":"tsa.connection.forward-worker.heartbeat.start","log_level":1,"data":{"remote":"10.0.84.97:44257","session":"15.1.22","worker-address":"0.0.0.0:33026","worker-platform":"linux","worker-tags":""}}
{"timestamp":"1477020577.976804972","source":"tsa","message":"tsa.connection.forward-worker.heartbeat.reached-worker","log_level":0,"data":{"remote":"10.0.84.97:44257","session":"15.1.22","took":"5.685945ms"}}
{"timestamp":"1477020577.984483004","source":"tsa","message":"tsa.connection.forward-worker.heartbeat.done","log_level":1,"data":{"remote":"10.0.84.97:44257","session":"15.1.22","worker-address":"0.0.0.0:33026","worker-platform":"linux","worker-tags":""}}
{"timestamp":"1477020578.038114548","source":"atc","message":"atc.garden-connection.retryable-http-client.retrying","log_level":1,"data":{"error":"dial tcp 0.0.0.0:33647: getsockopt: connection refused","failed-attempts":1,"ran-for":"179.927µs","session":"68203.1"}}
{"timestamp":"1477020578.360024214","source":"atc","message":"atc.garden-connection.retryable-http-client.retrying","log_level":1,"data":{"error":"dial tcp 0.0.0.0:33647: getsockopt: connection refused","failed-attempts":1,"ran-for":"172.181µs","session":"68204.1"}}
{"timestamp":"1477020579.044545650","source":"atc","message":"atc.garden-connection.retryable-http-client.retrying","log_level":1,"data":{"error":"dial tcp 0.0.0.0:33647: getsockopt: connection refused","failed-attempts":2,"ran-for":"1.00661154s","session":"68203.1"}}
{"timestamp":"1477020579.362408161","source":"atc","message":"atc.garden-connection.retryable-http-client.retrying","log_level":1,"data":{"error":"dial tcp 0.0.0.0:33647: getsockopt: connection refused","failed-attempts":2,"ran-for":"1.002554606s","session":"68204.1"}}
{"timestamp":"1477020579.436351299","source":"atc","message":"atc.garden-connection.retryable-http-client.retrying","log_level":1,"data":{"error":"dial tcp 0.0.0.0:33647: getsockopt: connection refused","failed-attempts":4,"ran-for":"7.014820109s","session":"68192.1"}}
```

**Comment by shinji62** [comment]
I think there is still no graceful stop of worker.
So maybe that the missing part which will allow autoscaling worker.

**Comment by vito** [comment]
This should be fixed with #629
